### PR TITLE
Add support for explicit package conflicts.

### PR DESCRIPTION
**Pull request by johntyree**
This adds explicit package conflicts to scenario files and rule generation. :dancer:

The format is just as with dependencies and it fully supports inequalities, e.g.

```
browsey 1.2.0-3;
    depends (click >= 1.1);
    conflicts (click <= 2.0, click > 1.3, zope ^= 0.19.3)
```

It includes unit tests and a test scenario file. If there's another conflicting scenario that should be included in here, please share it.

@cournape

### CODE REVIEW

**Comment by johntyree** [comment]
Note to self: Make sure that for conflicting packages, the removal occurs before the installation.

**Comment by johntyree** [comment]
Thinking about this, how consistent are we trying to be? Do we care about the intermediate states other than for conflicting packages?

Is it enough to always do all removals before all installations?

**Comment by johntyree** [comment]
We could be smarter about it by, say, finding all groups of conflicting packages and only enforcing the "remove before install" rule within the groups, but why bother?

**Comment by cournape** [comment]
@johntyree if possible, it is better to install dependencies first.

Right now, it is not a big deal, but later we may have some packages whose pre/post install scripts may depend on their dependencies. I would like to keep that option available, unless it is too complicated.

**Comment by johntyree** [comment]
Sure, we can do that. I just mean splitting things like package updates into a removal and later installation.

I already have the toposort code from the priority queue branch. I'll just try it out and see.

**Comment by johntyree** [comment]
I think if you want to maintain the invariant that we never try to install or remove a package without its dependencies being installed then we have to do away with the notion of "update" internally.

Currently we have a lot of things that differentiate somehow between "Remove a-0, install a-1" vs "update a from 0 to 1," but if we need to have b-0 installed while a-0 is being uninstalled, and b-1 installed while a-1 is being installed, then we have to split that operation up. From a UI perspective, I understand wanting to know which packages are being updated, but internally I think it should be simplified such that there are only `install` and `remove` operations.

The code for this is pretty straight forward. I have a working implementation already, that is contained entirely with the `Transaction` object. We should probably discuss a bit about how to go forward.

**Comment by johntyree** [comment]
There will be a lot of failing tests because of the removal of "update".

File: simplesat/transaction.py [hunk 1]
```diff
@@ -29,16 +30,20 @@ class RemoveOperation(Operation):
 class Transaction(object):
 
     def __init__(self, pool, decisions, installed_map):
-        self.operations = []
-
+        self.operations = self._safe_operations(pool, decisions, installed_map)
+        self.pretty_operations = []
```
Comment (johntyree) [inline simplesat/transaction.py]: The other code is still here for easy comparison. If we decide to do things the new way it should be removed.

**Comment by johntyree** [comment]
@cournape We can talk about this at our Monday meeting. Short of updating the tests, this is good to go.

We should decide if we want to keep the pretty representation with explicit `update` operations to show users. I'm leaning towards "no," so that we only show the user what we actually plan to do, but it depends on whether or not you think we're showing things "in order" or not in the first place.

If we say "update numpy, remove scipy, install seaborn" does that imply that we're going to do those things in that order? If so, then I propose we dump updates entirely. If not, then maybe we can keep them.

File: simplesat/tests/pil_pillow_upgrade_conflict.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+packages:
```
Comment (cournape) [inline simplesat/tests/pil_pillow_upgrade_conflict.yaml]: Are the PIL/Pillow-related scenario used anywhere ?

File: simplesat/transaction.py [hunk 3]
```diff
@@ -73,6 +78,44 @@ def to_simple_string(self):
...
+        for group in toposort(graph):
+            # Sort the set of independent packages for determinism
+            for package_id in sorted(group):
+                assert package_id in decisions
+                if package_id < 0 and -package_id in installed_map:
+                    removals.append(-package_id)
+                elif package_id > 0 and package_id not in installed_map:
+                    installs.append(package_id)
+
+        # Removals should happen top down
```
Comment (cournape) [inline simplesat/transaction.py]: Can we have some explicit unit test for this feature ?

File: simplesat/transaction.py [hunk 4]
```diff
@@ -73,6 +78,44 @@ def to_simple_string(self):
 
         return "\n".join(lines)
 
+    def install(self, package):
```
Comment (cournape) [inline simplesat/transaction.py]: This was there before this PR, but I don't think install/remove/update should be public APIs. Can we rename them to `_*` ?

File: simplesat/utils/graph.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python
```
Comment (cournape) [inline simplesat/utils/graph.py]: same here, those needs tests :)

File: simplesat/transaction.py (hunk 1 above)
Comment (cournape) [inline simplesat/transaction.py]: I think we should remove the `pretty_operations` computation (and all the related code in `_compute_transaction`), but in this case, the `Transaction` object should take care of generating nice update operations from the `operations` attribute.

**Comment by cournape** [comment]
@johntyree can you give an ETA to get this in a state where it can be merged in master ?

**Comment by johntyree** [comment]
@cournape this is today's project. It should be ready for re-review by end of day.

File: simplesat/transaction.py (hunk 1 above)
Comment (johntyree) [inline simplesat/transaction.py]: Well that took _forever_.

Should be all set now.

File: simplesat/tests/pil_pillow_upgrade_conflict.yaml (hunk 2 above)
Comment (johntyree) [inline simplesat/tests/pil_pillow_upgrade_conflict.yaml]: They are now.

File: simplesat/transaction.py (hunk 3 above)
Comment (johntyree) [inline simplesat/transaction.py]: Added tests for `Transaction` and the `graph` module.

File: simplesat/transaction.py (hunk 4 above)
Comment (johntyree) [inline simplesat/transaction.py]: These have been removed entirely.

File: simplesat/utils/graph.py (hunk 5 above)
Comment (johntyree) [inline simplesat/utils/graph.py]: :+1:

**Comment by cournape** [comment]
LGTM

**Pull request merged**